lib/gc/Transforms/Microkernel/ConvertLinalgToMicrokernel.cpp

```diff
@@ -0,0 +1,370 @@
+                          << ", B minor dim: " << brgemmDims.minorDimB << "\n");
+  return brgemmDims;
+}
+
+template <typename SrcBrmmOpTy>
+static FailureOr<linalg::TransposeOp> getFusibleTranspose(SrcBrmmOpTy brmmOp,
+                                                          Value buffer) {
+  auto defOp = buffer.getDefiningOp();
+  auto transOp = dyn_cast_or_null<linalg::TransposeOp>(defOp);
+  if (!transOp)
+    return failure();
+
+  using one_t = std::integral_constant<size_t, 1>;
+  using two_t = std::integral_constant<size_t, 2>;
+  constexpr size_t lastDimOffsetA = 1;
+  constexpr size_t lastDimOffsetB = std::conditional<
+      std::is_same<SrcBrmmOpTy, linalg::BatchReduceMatmulOp>::value, one_t,
+      two_t>::type::value;
+  size_t lastDimOffset =
+      buffer == brmmOp.getInputs()[0] ? lastDimOffsetA : lastDimOffsetB;
+
+  ArrayRef<int64_t> permutation = transOp.getPermutation();
+  bool lastDimContigious = true;
+  // Last dim can't not be permuted if we want to incorporate the
+  // transpose, because BRGEMM requires last dim to be contigious.
+  // For VNNI, it requires the last two dims to be non-permutedi
+  for (size_t idx = permutation.size() - lastDimOffset;
+       idx < permutation.size(); idx++)
+    lastDimContigious = lastDimContigious && (permutation[idx] == long(idx));
+
+  if (lastDimContigious)
+    return transOp;
+  return failure();
+}
+
+// Replace linalgOp with corresponding microkernel Op
+static void replaceOpWithMicrokernelOp(PatternRewriter &rewriter,
+                                       linalg::LinalgOp linalgOp,
+                                       const BrgemmDims &dims,
+                                       const DenseMap<Value, Value> &replaceMap,
+                                       bool isInitOutput) {
+  OpBuilder::InsertionGuard guard(rewriter);
+
+  DenseI64ArrayAttr batchDims = DenseI64ArrayAttr::get(
+      rewriter.getContext(), ArrayRef<int64_t>{dims.batchDimA, dims.batchDimB});
+  DenseI64ArrayAttr leadingDims = DenseI64ArrayAttr::get(
+      rewriter.getContext(),
+      ArrayRef<int64_t>{dims.leadingDimA, dims.leadingDimB});
+
+  SmallVector<Attribute> brgemmFlags;
+  if (isInitOutput) {
+    brgemmFlags.push_back(microkernel::BrgemmFlagsAttr::get(
+        rewriter.getContext(), microkernel::BrgemmFlags::BETA_0));
+  }
+  auto flags = rewriter.getArrayAttr(brgemmFlags);
+
+  Value operandA = linalgOp.getDpsInputOperands()[0]->get();
+  Value operandB = linalgOp.getDpsInputOperands()[1]->get();
+  Value operandC = linalgOp.getDpsInitsMutable()[0].get();
+
+  SmallVector<Value> inputs{operandA, operandB};
+  auto brgemmOp = rewriter.replaceOpWithNewOp<microkernel::BrgemmOp>(
+      linalgOp, operandC.getType(), inputs, operandC, batchDims, leadingDims,
+      flags);
+  // Replace operands according to fusion
+  rewriter.modifyOpInPlace(brgemmOp, [&]() {
+    for (const auto &pair : replaceMap)
+      brgemmOp->replaceUsesOfWith(pair.first, pair.second);
```

Comment:
The `pair.first` should also check there is no other use except brgemm

Reply:
Here we only replace use of `brgemmOp`, and `pair.first` (output of `transposeOp`) is not necessarily used by `brgemmOp` only . If it's only use, then `transpose` would be eliminated later as it's dead code; else `transpose` would be preserved and basically nothing major happens.